Rework PyObject preservation (v2)

colesbury opened the pull request:
Make the PyObject preservation scheme thread-safe with free threaded (nogil) Python. The general idea is:

* Python Tensor and Storage objects always hold a strong reference to their underlying c10 object
* c10 objects hold a strong reference to their Python objects if there's at least one other reference to the c10 object

This is implemented in `intrusive_ptr`:

* The top most bit (`kHasPyObject`) from the weakref count is now used to indicate if the `intrusive_ptr_target` has an associated PyObject. So `kHasPyObject` is one bit, the weakref count is now 31 bits and the strong refcount remains 32 bits.
* When the reference count increases from one to two and `kHasPyObject` is set, we incref the associated Python object to ensure that it's kept alive. 
* When the reference count decreases from two to one (i.e., there are no C++ reference to the `intrusive_ptr_target` other than from the Python object), we decre the associated Python object to break the cycle.

Other benefits:

* We can delete a lot of the copypasta from Python internal `subtype_dealloc`
* This fixes the weakref and GC bugs we had in the previous scheme. Python weakrefs on Tensors and Storages should just work as expected now.

Risks:

* Extra branch for reference count operations on `intrusive_ptr<TensorImpl>`, `intrusive_ptr<StorageImpl>`, and the generic `intrusive_ptr<intrusive_ptr_target>` even when we're not using Python.
* It's a big change

(Second attempt at https://github.com/pytorch/pytorch/pull/166342)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @albanD

albanD reviewed (approved):
Sounds good if this fixes CI

Skylion007 commented on c10/core/impl/PyObjectSlot.h:
```diff
@@ -8,117 +8,70 @@
-  void set_owns_pyobj(bool b);
+  // Non thread-safe swap
+  void swap(PyObjectSlot& other) noexcept {
+    PyInterpreter* tmp_interpreter =
```
Can this be simplified using std::atomic exchange somehow? Feels like that would do the exact same thing with way less temporaries and way cleaner. Also would be more efficient with regards to atomic locks / updates I think.

colesbury replied:
Yeah, that's `std::atomic<T>::exchange`, which is (relatively) expensive, and different from `std::exchange`:

* https://en.cppreference.com/w/cpp/utility/exchange.html
* https://en.cppreference.com/w/cpp/atomic/atomic/exchange.html

Performance doesn't matter much here because it's only used by `THPModule_swap_tensor_impl`, but I think things will be cleaner if we:

1) Remove PyObjectSlot::swap()
2) Add PyObjectSlot::store_pyobj()
3) Do whatever minimal swapping logic in `THPModule_swap_tensor_impl` using `PyObjectSlot::store_pyobj()`

Skylion007 reviewed (approved):
Unblocking for now, please update with std::exchange changes in a follow up PR

colesbury committed:
Rework PyObject preservation (v2) (#167564)

Summary:
Make the PyObject preservation scheme thread-safe with free threaded (nogil) Python. The general idea is:

* Python Tensor and Storage objects always hold a strong reference to their underlying c10 object
* c10 objects hold a strong reference to their Python objects if there's at least one other reference to the c10 object

This is implemented in `intrusive_ptr`:

* The top most bit (`kHasPyObject`) from the weakref count is now used to indicate if the `intrusive_ptr_target` has an associated PyObject. So `kHasPyObject` is one bit, the weakref count is now 31 bits and the strong refcount remains 32 bits.
* When the reference count increases from one to two and `kHasPyObject` is set, we incref the associated Python object to ensure that it's kept alive.
* When the reference count decreases from two to one (i.e., there are no C++ reference to the `intrusive_ptr_target` other than from the Python object), we decre the associated Python object to break the cycle.

Other benefits:

* We can delete a lot of the copypasta from Python internal `subtype_dealloc`
* This fixes the weakref and GC bugs we had in the previous scheme. Python weakrefs on Tensors and Storages should just work as expected now.

Risks:

* Extra branch for reference count operations on `intrusive_ptr<TensorImpl>`, `intrusive_ptr<StorageImpl>`, and the generic `intrusive_ptr<intrusive_ptr_target>` even when we're not using Python.
* It's a big change

(Second attempt at https://github.com/pytorch/pytorch/pull/166342)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben albanD


Differential Revision: D86936370

Pulled By: colesbury

rgommers commented:
Reverting based on that one report does not seem reasonable. This PR is quite important and fixes multiple bugs (as can be seen from all the other incoming links above). 

The report in gh-170743 has been shown to be bottlenecked on the GIL due to use of Python-level threading, which is always going to see some changes in performance as fixes for thread-safety are made. There are no other reports of performance impact in the 2 months after this PR was merged.
